R/utils.R

```diff
@@ -92,3 +92,123 @@ browse_url <- function(path, chromote) {
     utils::browseURL(url)
   }
 }
+
+
+# =============================================================================
+# Random Ports
+# =============================================================================
+#
+# Borrowed from https://github.com/rstudio/httpuv/blob/main/R/random_port.R
+
+with_random_port <- function(
+  startup,
+  min = 1024L,
+  max = 49151L,
+  n = 5,
+  host = "http://127.0.0.1"
+) {
+  stopifnot(is.function(startup))
+  valid_ports <- setdiff(seq.int(min, max), unsafe_ports)
+
+  # Try up to n ports
+  n <- min(n, length(valid_ports))
+  ports <- sample(valid_ports, n)
+  for (port in ports) {
+    res <- NULL
+
+    # Try to run `.f()` with the random port
+    res <- tryCatch(
+      startup(host = host, port = port),
+      error_timeout = identity,
+      system_command_error = identity,
+      error = function(err) if (identical(port, ports[n])) err
+    )
+
+    if (rlang::cnd_inherits(res, "error")) {
+      # A timeout error or trying all ports is unlikely to be port-related problem
+      rlang::cnd_signal(res)
+    }
+
+    if (!is.null(res)) {
+      return(res)
+    }
+  }
+
+  rlang::abort(
+    "Cannot find an available port. Please try again.",
+    class = "chromote_error_no_available_port"
+  )
```

Comment:
I believe that with the existing code, it will never reach this code, because it would throw an error at line 129 above.

But with the revised version I suggested above, it would get here.

Reply:
Good catch, thank you! My intention was to include the last port-failure error in our final error. I added a separate variable `err_port` that tracks the last port error so that we can include it in the error chain of the final error if we exhaust all port attempts.